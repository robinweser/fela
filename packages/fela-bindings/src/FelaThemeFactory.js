/* @flow */
import { THEME_CHANNEL } from './themeChannel'

export default function FelaThemeFactory(
  BaseComponent: any,
  contextTypes?: Object
): any {
  class FelaTheme extends BaseComponent {
    unsubscribe: ?Function
    constructor(props, context) {
      super(props, context)

      this.state = {
        theme: context[THEME_CHANNEL] ? context[THEME_CHANNEL].get() : {},
      }
    }

    componentWillMount() {
      if (this.context[THEME_CHANNEL]) {
        this.unsubscribe = this.context[THEME_CHANNEL].subscribe(properties =>
          this.setState({
            theme: properties,
          })
        )
      }
    }

    componentWillUnmount() {
      if (this.unsubscribe) {
        this.unsubscribe()
      }
    }

    render() {
      return this.props.render(this.state.theme)
    }
  }

  if (contextTypes) {
    FelaTheme.contextTypes = contextTypes
  }

  return FelaTheme
}
