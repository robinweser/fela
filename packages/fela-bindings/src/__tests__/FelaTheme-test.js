import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import FelaThemeFactory from '../FelaThemeFactory'
import createTheme from '../createTheme'
import { THEME_CHANNEL } from '../themeChannel'

const FelaTheme = FelaThemeFactory(Component, {
  [THEME_CHANNEL]: PropTypes.object,
})

describe('Using the FelaTheme component', () => {
  it('correctly pass the theme down', () => {
    const themeContext = createTheme({
      color: 'red',
    })

    const wrapper = mount(
      <FelaTheme render={theme => <div>The color is {theme.color}.</div>} />,
      {
        context: {
          [THEME_CHANNEL]: themeContext,
        },
      }
    )

    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
