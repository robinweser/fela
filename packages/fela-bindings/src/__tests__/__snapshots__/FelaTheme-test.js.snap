// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using the FelaTheme component correctly pass the theme down 1`] = `
<FelaTheme
  render={[Function]}
>
  <div>
    The color is 
    red
    .
  </div>
</FelaTheme>
`;
